File: integrations/harmony/HarmonyWebsocket.go
package harmony

import (
	"bytes"
	"encoding/json"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"net/url"
	"strconv"
	"sync"
	"time"

	"github.com/gorilla/websocket"
	"github.com/insomniadev/martian/database"
	"github.com/insomniadev/martian/modules/pubsub"
)

var (
	defaultHubPort = 8088
)

// Init the harmony hub communication
func (d *Device) Init(configuration string) error {
	// Load up the configuration as the device itself
	json.Unmarshal([]byte(configuration), d)

	// IF the IPAddress doesn't exist, then we need to find the endpoint
	if d.IPAddress == "" {
		d.discover()
	}

	d.connect()
	err := database.MartianData.PutIntegrationValue("harmony", d)
	if err != nil {
		return err
	}
	return nil
}

// discover will find the harmony device on the network
func (d *Device) discover() {
	addr := "192.168.1.1/24"
	ip, ipnet, err := net.ParseCIDR(addr)
	if err != nil {
		panic(err)
	}
	var wg sync.WaitGroup
	for ip := ip.Mask(ipnet.Mask); ipnet.Contains(ip); inc(ip) {
		wg.Add(1)
		go func(ipdad string) {
			defer wg.Done()
			message, err := checkIpAddress(ipdad)
			if err != nil {
				return
			}
			d.IPAddress = ipdad + ":" + strconv.Itoa(defaultHubPort)
			d.ActiveRemoteID = message.Data.ActiveRemoteID
			u, err := url.Parse(message.Data.DiscoveryServer)
			if err != nil {
				log.Fatal(err)
			}
			d.HostName = u.Hostname()

		}(ip.String())
	}
	wg.Wait()
}
func inc(ip net.IP) {
	for j := len(ip) - 1; j >= 0; j-- {
		ip[j]++
		if ip[j] > 0 {
			break
		}
	}
}
func checkIpAddress(ipAddress string) (HTTPMessage, error) {
	host := "http://" + ipAddress + ":" + strconv.Itoa(defaultHubPort)
	jsonBody := []byte(`{"id":1,"cmd":"setup.account?getProvisionInfo","params":{}}`)
	client := &http.Client{Timeout: 500 * time.Millisecond}
	req, err := http.NewRequest("Post", host, bytes.NewBuffer(jsonBody))
	if err != nil {
		log.Fatal(err)
	}
	req.Header.Add("Origin", "http://sl.dhg.myharmony.com")
	req.Header.Add("Content-Type", "application/json")
	req.Header.Add("Accept", "application/json")
	req.Header.Add("Accept-Charset", "utf-8")
	resp, err := client.Do(req)
	if err != nil {
		return HTTPMessage{}, err
	}

	defer resp.Body.Close()

	body, _ := ioutil.ReadAll(resp.Body)
	message := HTTPMessage{}
	if err = json.Unmarshal(body, &message); err != nil {
		log.Fatal(err)
		return HTTPMessage{}, err
	}
	return message, nil
}

func (d *Device) connect() {
	host := "ws://" + d.IPAddress + "/?domain=" + d.HostName + "&hubId=" + strconv.Itoa(d.ActiveRemoteID)

	conn, _, err := websocket.DefaultDialer.Dial(host, nil)
	d.Connection = conn
	if err != nil {
		log.Fatal("harmony dial:", err)
	}
	go d.listen()
	d.GetConfig()
	d.GetCurrentActivity()
}

// heartbeat sends a websocket request every 30 seconds so that the websocket connection stays open
func (d *Device) heartbeat() {
	for {
		time.Sleep(30 * time.Second)
		d.WriteMessage("")
	}
}

func (d *Device) listen() {
	go d.heartbeat()
	for {
		time.Sleep(1 * time.Second)
		_, message, err := d.Connection.ReadMessage()
		if err != nil {
			log.Println("harmony read:", err.Error())
			time.Sleep(30 * time.Second) // Wait for thirty seconds
			go d.connect()               // Start a new process to connect
			return                       // exit out of this current listening loop
		}
		if len(string(message)) > 0 {
			receivedMessage := RecvMessage{}
			err = json.Unmarshal(message, &receivedMessage)
			if receivedMessage.ID == 1 {
				receivedResult := RecvResult{}
				err = json.Unmarshal(message, &receivedResult)
				d.CurrentActivity = receivedResult.Data.Result
				d.ActivityID = receivedResult.Data.Result
			} else if receivedMessage.ID == 2 {
				receivedResult := RecvConfig{}
				err = json.Unmarshal(message, &receivedResult)
				activitiesRecorded := make([]Activity, 0)
				for _, data := range receivedResult.Data.Activity {
					deviceActions := make([]ActivityActions, 0)
					for _, device := range data.ControlGroup {
						for _, funct := range device.Function {
							actCommand := ActivityActionCommand{
								Command:  funct.Action,
								DeviceID: funct.Name,
								Type:     funct.Label,
							}
							actAction := ActivityActions{
								Label:  device.Name,
								Action: actCommand,
							}
							deviceActions = append(deviceActions, actAction)
						}
					}
					nextActivity := Activity{
						ActivityID: data.ID,
						Name:       data.Label,
						Actions:    deviceActions,
					}
					activitiesRecorded = append(activitiesRecorded, nextActivity)
				}
				d.Activities = activitiesRecorded

				database.MartianData.PutIntegrationValue("harmony", d)
			} else if receivedMessage.Type == "connect.stateDigest?notify" {
				receivedResult := StateDigestNotify{}
				err = json.Unmarshal(message, &receivedResult)
				d.CurrentActivity = receivedResult.Data.ActivityID
				// FIXME: Fix the brain notification
				formattedString := "harmony;;harmony;;" + d.CurrentActivity
				pubsub.Service.Publish("brain", formattedString)
				pubsub.Service.Publish("subscriptions", d.CurrentActivity)
			}
		}
	}
}

// GetCurrentActivity calls the websocket to return the current activity
func (d *Device) GetCurrentActivity() {
	message := `{"hubId":"` + strconv.Itoa(d.ActiveRemoteID) + `","timeout":30, "hbus":{"cmd":"vnd.logitech.harmony/vnd.logitech.harmony.engine?getCurrentActivity","id":1,"params":{"verb":"get","format":"json"}}}`
	d.WriteMessage(message)
}

// GetConfig calls the websocket to return the hub configuration
func (d *Device) GetConfig() {
	message := `{"hubId":"` + strconv.Itoa(d.ActiveRemoteID) + `","timeout":30, "hbus":{"cmd":"vnd.logitech.harmony/vnd.logitech.harmony.engine?config","id":2,"params":{"verb":"get","format":"json"}}}`
	d.WriteMessage(message)
}

// StartActivity starts an activity
func (d *Device) StartActivity(activityID string) {
	message := `{"hubId":"` + strconv.Itoa(d.ActiveRemoteID) + `","timeout":30, "hbus":{"cmd":"harmony.activityengine?runactivity","id":3,"params":{"async":"true","timestamp":"0","args":{"rule":"start"},"activityId":"` + activityID + `"}}}`
	d.WriteMessage(message)
}

// SendCommand sends a command to the harmony hub
func (d *Device) SendCommand(command string, commandType string, deviceID string) {
	nextCommand := &Command{
		Command:  command,
		Type:     commandType,
		DeviceID: deviceID,
	}
	d.sendCommandToHub(*nextCommand)
}

// sendCommandToHub sends a command to the harmony hub
func (d *Device) sendCommandToHub(action Command) {
	actionJSON, err := json.Marshal(action)
	if err != nil {
		log.Println("Harmony Command Error:", err)
	}
	message := `{"hubId":` + strconv.Itoa(d.ActiveRemoteID) + `,"timeout":30, "hbus":{"cmd":"vnd.logitech.harmony/vnd.logitech.harmony.engine?holdAction","id":4,"params":{"status":"pressrelease","timestamp":"0","verb":"render","action":"` + string(actionJSON) + `"}}}`
	d.WriteMessage(message)
}

// WriteMessage will write a message to the harmony websocket
func (d *Device) WriteMessage(message string) {
	err := d.Connection.WriteMessage(websocket.TextMessage, []byte(message))
	if err != nil {
		log.Println("harmony write:", err)
	}
}

// EditDeviceConfiguration will update the name and areaname for the harmony configuration
func (d *Device) EditDeviceConfiguration(device Device, removeEdit bool) {
	d.Name = device.Name
	d.AreaName = device.AreaName

	// Save in the database
	err := database.MartianData.PutIntegrationValue("harmony", d)
	if err != nil {
		log.Println(err)
	}
}

File: integrations/harmony/harmonyWebSocket.js
import { Config } from "../../src/models/config";
import EventEmitter from "events";
import request from "request-promise";
import { w3cwebsocket as WebSocketClient } from "websocket";
import WebSocketAsPromised from "websocket-as-promised";

const Url = require("url");
const DEFAULT_HUB_PORT = 8088;

const HUB_CONNECT_TIMEOUT = 10000;
const HUB_SEND_TIMEOUT = 30000;

export class HarmonyWebSocket extends EventEmitter {
  constructor() {
    super();
    const { ipAddress } = new Config().getHarmonyConfig();
    this.ipAddress = ipAddress;
    super();
    this._remoteId = null;
    this._domain = null;
    this._client = null;
    this._interval = null;
    this._connectTimeout = HUB_CONNECT_TIMEOUT;
    this._sendTimeout = HUB_SEND_TIMEOUT;
  }

  set connectTimeout(timeout) {
    this._connectTimeout = timeout;
  }

  set sendTimeout(timeout) {
    this._sendTimeout = timeout;
  }

  connect() {
    return this._getRemoteId()
      .then((data) => {
        this._remoteId = data.data.activeRemoteId;

        this._domain = Url.parse(data.data.discoveryServer).hostname;
      })
      .then(() => this._connect());
  }

  isOpened() {
    return this._client ? this._client.isOpened : false;
  }

  close() {
    clearInterval(this._interval);
    this.emit("close");
    return this.isOpened() ? this._client.close() : false;
  }

  _getRemoteId() {
    const hubUrl = `http://${this.ipAddress}:${DEFAULT_HUB_PORT}/`;

    const headers = {
      Origin: "http://sl.dhg.myharmony.com",
      "Content-Type": "application/json",
      Accept: "application/json",
      "Accept-Charset": "utf-8",
    };

    const jsonBody = {
      id: 1,
      cmd: "setup.account?getProvisionInfo",
      params: {},
    };

    const payload = {
      url: hubUrl,
      method: "POST",
      timeout: this._connectTimeout,
      headers: headers,
      json: true,
      body: jsonBody,
    };

    return request(payload);
  }

  _connect() {
    const url = `ws://${this.ipAddress}:${DEFAULT_HUB_PORT}/?domain=${this._domain}&hubId=${this._remoteId}`;

    this._client = new WebSocketAsPromised(url, {
      createWebSocket: (url) => new WebSocketClient(url),
      packMessage: (data) => JSON.stringify(data),
      unpackMessage: (message) => JSON.parse(message),
      attachRequestId: (data, requestId) => {
        data.hbus.id = requestId;
        return data;
      },
      extractRequestId: (data) => data && data.id,
      connectionTimeout: this._connectTimeout,
      timeout: this._timeout,
    });

    this._client.onClose.addListener((data) => {
      clearInterval(this._interval);
      console.log(JSON.stringify(data));
      this.emit("close");
    });

    this._client.onOpen.addListener(() => {
      clearInterval(this._interval);
      this._heartbeat();
      this.emit("open");
    });

    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "vnd.logitech.connect/vnd.logitech.statedigest?get",
        id: 0,
        params: {
          verb: "get",
          format: "json",
        },
      },
    };

    return this._client
      .open()
      .then(() =>
        this._client.onUnpackedMessage.addListener(this._onMessage.bind(this))
      )
      .then(() => this._client.sendPacked(payload));
  }

  _heartbeat() {
    // timeout = 60s
    this._interval = setInterval(() => {
      try {
        this._client.send("");
      } catch (e) {
        this.close();
      }
    }, 50000);
  }

  _onMessage(message) {
    if (message.type === "connect.stateDigest?notify") {
      this.emit("stateDigest", message);
    } else if (message.type === "automation.state?notify") {
      this.emit("automationState", message);
    }
  }

  getCapabilities() {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "proxy.resource?get",
        id: 0,
        params: {
          uri: `harmony://Account/${this._remoteId}/CapabilityList`,
        },
      },
    };

    return this._client.open().then(() => this._client.sendRequest(payload));
  }

  getConfig() {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "vnd.logitech.harmony/vnd.logitech.harmony.engine?config",
        id: 0,
        params: {
          verb: "get",
          format: "json",
        },
      },
    };

    return this._client.open().then(() => this._client.sendRequest(payload));
  }

  getAutomationConfig() {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "proxy.resource?get",
        id: 0,
        params: {
          uri: "dynamite://HomeAutomationService/Config/",
        },
      },
    };

    return this._client.open().then(() => this._client.sendRequest(payload));
  }

  getActivities() {
    return this.getConfig().then((response) => {
      return response.data.activity.map((action) => {
        return {
          id: action.id,
          label: action.label,
        };
      });
    });
  }

  getCurrentActivity() {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd:
          "vnd.logitech.harmony/vnd.logitech.harmony.engine?getCurrentActivity",
        id: 0,
        params: {
          verb: "get",
          format: "json",
        },
      },
    };

    return this._client
      .open()
      .then(() => this._client.sendRequest(payload))
      .then((response) => response.data.result);
  }

  startActivity(activityId) {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "harmony.activityengine?runactivity",
        id: 0,
        params: {
          async: "true",
          timestamp: 0,
          args: {
            rule: "start",
          },
          activityId: activityId,
        },
      },
    };

    return this._client
      .open()
      .then(() => this._client.sendRequest(payload))
      .then((response) => response);
  }

  getActivityCommands(activityId) {
    if (activityId === "-1") return;
    return this.getConfig().then((response) => {
      var activity = response.data.activity
        .filter((act) => {
          return act.id === activityId;
        })
        .pop();
      return activity.controlGroup
        .map((group) => {
          return group.function;
        })
        .reduce((prev, curr) => {
          return prev.concat(curr);
        })
        .map((fn) => {
          return {
            action: JSON.parse(fn.action),
            label: fn.label,
          };
        });
    });
  }

  getDevices() {
    return this.getConfig().then((response) => {
      return response.data.device
        .filter((device) => {
          return device.controlGroup.length > 0;
        })
        .map((device) => {
          return {
            id: device.id,
            label: device.label,
          };
        });
    });
  }

  getDeviceCommands(deviceId) {
    return this.getConfig().then((response) => {
      var device = response.data.device
        .filter((device) => {
          return device.id === deviceId;
        })
        .pop();
      return device.controlGroup
        .map((group) => {
          return group.function;
        })
        .reduce((prev, curr) => {
          return prev.concat(curr);
        })
        .map((fn) => {
          return {
            action: JSON.parse(fn.action),
            label: fn.label,
          };
        });
    });
  }

  getAutomationCommands() {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "harmony.automation?getstate",
        id: 0,
        params: {
          format: "json",
          forceUpdate: true,
        },
      },
    };

    return this._client.open().then(() => this._client.sendRequest(payload));
  }

  sendCommandWithDelay(action, hold) {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "vnd.logitech.harmony/vnd.logitech.harmony.engine?holdAction",
        id: 0,
        params: {
          status: "press",
          timestamp: "0",
          verb: "render",
          action: action,
        },
      },
    };

    return this._client
      .open()
      .then(() => {
        this._client.sendPacked(payload);
      })
      .then(() => {
        payload.hbus.params.status = "release";
        payload.hbus.params.timestamp = hold.toString();
      })
      .then(() => {
        this._client.sendPacked(payload);
      })
      .then(() =>
        Promise.resolve({
          cmd: payload.hbus.cmd,
          code: 200,
          id: payload.hbus.id,
          msg: "OK",
        })
      );
  }

  sendCommand(action) {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "vnd.logitech.harmony/vnd.logitech.harmony.engine?holdAction",
        id: 0,
        params: {
          status: "pressrelease",
          timestamp: "0",
          verb: "render",
          action: action,
        },
      },
    };

    return this._client.open().then(() => this._client.sendRequest(payload));
  }

  sendAutomationCommand(action) {
    var payload = {
      hubId: this._remoteId,
      timeout: 30,
      hbus: {
        cmd: "harmony.automation?setstate",
        id: 0,
        params: {
          state: {},
        },
      },
    };

    Object.assign(payload.hbus.params.state, action);

    return this._client.open().then(() => this._client.sendRequest(payload));
  }
}